### PR TITLE
Refactor WebGL implementation

### PR DESCRIPTION
**Pull request by emilio**
This PR uses customized GL context creation code, right now only working under Linux, so I expect the clearcolor test to fail on other platforms.

It addresses some other problems:
- Propagates context creation error to the top, returning null if not found.
- Uses `GLContextAttributes`, which will allow us to write `WebGLContextAttributes` easily.
- Doesn't allow a 2d context and a WebGL context coexist.
- Panics when resizing the context to larger dimensions (to be fixed soon, but better than blindly allowing it).
- Removes some unused dependencies

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5769)

### CODE REVIEW

**Comment by jdm** [comment]
Reviewed files:
- components/canvas/Cargo.toml @ r1
- components/canvas/lib.rs @ r1
- components/canvas/webgl_paint_task.rs @ r1
- components/script/dom/htmlcanvaselement.rs @ r1
- components/script/dom/webglrenderingcontext.rs @ r1
- components/script/dom/webidls/HTMLCanvasElement.webidl @ r1
- components/servo/Cargo.lock @ r1
- ports/cef/Cargo.lock @ r1
- ports/gonk/Cargo.lock @ r1
- tests/ref/webgl-context/clearcolor.html @ r1

---

<sup>**[components/canvas/webgl_paint_task.rs, line 91 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQt-Dw6ILS2zqNsLCi)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/canvas/webgl_paint_task.rs#L91)):</sup>
nit: || on the previous line.

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtEdkbaxySrrCR1IS)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L174)):</sup>
Use debug! instead of println!

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 190 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQu1nqJCIi90Uxolky)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L190)):</sup>
Why not use or_init here like before?

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 192 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtmGLPzVNwmfgGxKg)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L192)):</sup>
Why this change?

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 196 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtYCAz0Gc7CSWCbvP)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L196)):</sup>
This match can be a call to `self.context_webgl.get().map` instead.

---

<sup>**[components/script/dom/webglrenderingcontext.rs, line 43 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQuG5ATWfysf9MB5_p)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/webglrenderingcontext.rs#L43)):</sup>
Let's use error! instead of println!

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by dmarcos** [comment]
@jdm @ecoal95 What's the rational for these changes? What problem is the ```rust-offscreen-rendering-context``` solving? It's not glutin supposed to address the problem of creating and manage opengl contexts in a cross platform way? It already considers headless / offscreen contexts:

https://github.com/servo/glutin/blob/servo/tests/headless.rs

**Comment by ecoal95** [comment]
@dmarcos Hardware acceleration over all. Glutin uses OSMesa for offscreen rendering under the hood (see https://github.com/servo/servo/wiki/Meeting-2015-04-06).

It gives us more flexibility to implement WebGLContextOptions and all that stuff too.

**Comment by dmarcos** [comment]
@ecoal95 Can glutin be improved instead of having another library to manage OpenGL contexts? Can you for instance replace the use of OSMesa in glutin on linux? Is in the roadmap of glutin to move away of OSMesa? Can we live with a worse performance in the meantime and wait for glutin to get better? I think that at this stage is more valuable to keep servo working in a cross platform than getting hardware acceleration. I won't be able to contribute to WebGL if this stops working on OSX. Do you have plans to make your library work on the Mac?

**Comment by ecoal95** [comment]
<sup>**[components/script/dom/htmlcanvaselement.rs, line 190 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQu1nqJCIi90Uxolky)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L190)):</sup>
Because `WebGLRenderingContext#new` returns `None` if the context couldn't be created:
`or_init` takes a lambda and expects a return type of T, instead of `Option<T>`.

I can change it to return a `Result` instead and call `or_init` just if the call succeeds if you prefer.

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 192 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtmGLPzVNwmfgGxKg)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L192)):</sup>
Good catch!

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by ecoal95** [comment]
@dmarcos Yes, for sure, I plan to get it working in all platforms.

I don't think glutin plans to move away from OSMesa, or has the same need of customisation we may need.

Anyways, I agree with you... As of right now, what about conditionally falling back to glutin? Right now it should be easy, just changing the type of the gl_context value for linux/non-linux systems, or even falling back completely. The rest of the changes apply (instead of calling `unwrap()` on context creation).

What do you think @jdm?

**Comment by jdm** [comment]
Yes, I would much prefer falling back to cross-platform solutions if we don't support accelerated ones.

**Comment by jdm** [comment]
Reviewed files:
- components/canvas/webgl_paint_task.rs @ r2
- components/script/dom/htmlcanvaselement.rs @ r2
- components/script/dom/webglrenderingcontext.rs @ r2

---





Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by dmarcos** [comment]
@ecoal95 What about integrating your patch in glutin so we don't use OSMesa on Linux to get the hardware acceleration. Can we make glutin also more flexible so we can use WebGLContextOptions? Your library and glutin are redundant and it will introduce confusion in the long term. I think It's better to focus the energy on a single solution. @jdm What do you think?

File: components/canvas/webgl_paint_task.rs [hunk 1]
```diff
@@ -2,36 +2,44 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use canvas_msg::{CanvasWebGLMsg, CanvasCommonMsg, CanvasMsg};
+use canvas_msg::{CanvasMsg, CanvasWebGLMsg, CanvasCommonMsg};
```
Comment (dmarcos) [inline components/canvas/webgl_paint_task.rs]: This change is not needed

File: components/script/dom/webidls/HTMLCanvasElement.webidl [hunk 2]
```diff
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-//  https://www.whatwg.org/html/#htmlcanvaselement
+// http://www.whatwg.org/html/#htmlcanvaselement
```
Comment (dmarcos) [inline components/script/dom/webidls/HTMLCanvasElement.webidl]: This change is not needed

File: tests/ref/webgl-context/clearcolor.html [hunk 3]
```diff
@@ -14,12 +14,11 @@
 <body>
 <canvas id="canvas" width="640" height="480"></canvas>
 <script type="text/javascript">
-
-var gl = document.getElementById("canvas").getContext("webgl");
+var canvas = document.getElementById("canvas"),
```
Comment (dmarcos) [inline tests/ref/webgl-context/clearcolor.html]: Not needed

**Comment by jdm** [comment]
@dmarcos @glennw is our main glutin contributor and thinks this work is fine as-is. Glutin also tends to get ahead of Servo's rustc version, so for the purposes of @ecoal95's project I'm content to do the experimentation in a separate repository and see where we end up.

**Comment by ecoal95** [comment]
This should do the work for now :P

File: components/canvas/webgl_paint_task.rs (hunk 1 above)
Comment (ecoal95) [inline components/canvas/webgl_paint_task.rs]: Yep, all those subtle changes are from the rebase, they don't hurt anyways.

Comment (dmarcos) [inline components/canvas/webgl_paint_task.rs]: You never know and it makes the patch noisier and harder to review. The only free bug code is the code you never write.

**Comment by jdm** [comment]
Reviewed files:
- components/canvas/Cargo.toml @ r3
- components/canvas/lib.rs @ r3
- components/canvas/webgl_paint_task.rs @ r3
- components/script/dom/webidls/HTMLCanvasElement.webidl @ r3
- components/servo/Cargo.lock @ r3

---

<sup>**[components/servo/Cargo.lock, line 65 \[r3\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnRevXFclnuxFDqXxEy)** ([raw file](https://github.com/servo/servo/blob/f856558d38bdd52d6a5aa2c54029aeb5d6d2bfa6/components/servo/Cargo.lock#L65)):</sup>
Please also run `./mach build-cef` and `GONKDIR=/tmp ./mach build-gonk`; you can interrupt them as soon as you see "compiling" in the output, but they'll cause the other Cargo.lock files to update.

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by ecoal95** [comment]
<sup>**[components/canvas/webgl_paint_task.rs, line 91 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQt-Dw6ILS2zqNsLCi)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/canvas/webgl_paint_task.rs#L91)):</sup>
Done.

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtEdkbaxySrrCR1IS)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L174)):</sup>
Done.

---

<sup>**[components/script/dom/htmlcanvaselement.rs, line 196 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtYCAz0Gc7CSWCbvP)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L196)):</sup>
Done.

---

<sup>**[components/script/dom/webglrenderingcontext.rs, line 43 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQuG5ATWfysf9MB5_p)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/webglrenderingcontext.rs#L43)):</sup>
Done.

---

<sup>**[components/servo/Cargo.lock, line 65 \[r3\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnRevXFclnuxFDqXxEy)** ([raw file](https://github.com/servo/servo/blob/f856558d38bdd52d6a5aa2c54029aeb5d6d2bfa6/components/servo/Cargo.lock#L65)):</sup>
Done! I did it the first time, but I forgot this one, sorry :/

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by dmarcos** [comment]
@ecoal95 I'm testing your patch in OSX and getting the following compilation errors: https://pastebin.mozilla.org/8830856

**Comment by jdm** [comment]
<sup>**[components/script/dom/htmlcanvaselement.rs, line 174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtEdkbaxySrrCR1IS)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L174)):</sup>
You fixed the other one, but not this one :)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by ecoal95** [comment]
@dmarcos try now please :P

---









Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by ecoal95** [comment]
<sup>**[components/script/dom/htmlcanvaselement.rs, line 174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5769#-JnQtEdkbaxySrrCR1IS)** ([raw file](https://github.com/servo/servo/blob/ca8799037c3e2ab35a25d5f0f83b1b9896813dda/components/script/dom/htmlcanvaselement.rs#L174)):</sup>
Dang it! Done :P

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5769)

**Comment by jdm** [comment]
@larsbergstrom Here's another instance of reviewable not picking up a new commit.

**Comment by dmarcos** [comment]
@ecoal95 It builds and I see pixels when running the test ```./mach run tests/ref/webgl-context/clearcolor.html``` Thanks for keeping it working cross platform!

**Comment by ecoal95** [comment]
Cool then! :)

**Comment by jdm** [comment]
Ok, I think this is good to go. @ecoal95 Could you squash this all into one commit?

**Comment by ecoal95** [comment]
Done :P

**Comment by jdm** [comment]
```
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/draw_buffer.rs:98:61: 98:70 error: unresolved name `gl::RGBA8`
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/draw_buffer.rs:98             self.color_render_buffer = create_render_buffer(gl::RGBA8, &self.size);
                                                                                                                                                                            ^~~~~~~~~
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/draw_buffer.rs:100:61: 100:69 error: unresolved name `gl::RGB8`
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/draw_buffer.rs:100             self.color_render_buffer = create_render_buffer(gl::RGB8, &self.size);
                                                                                                                                                                             ^~~~~~~~
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/gl_context_capabilities.rs:23:38: 23:53 error: unresolved name `gl::MAX_SAMPLES`
/home/servo/.cargo/git/checkouts/rust-offscreen-rendering-context-902f85b061901132/master/src/gl_context_capabilities.rs:23             unsafe { gl::GetIntegerv(gl::MAX_SAMPLES, &mut capabilities.max_samples as *mut GLint); };
                                                                                                                                                                 ^~~~~~~~~~~~~~~
```

**Comment by ecoal95** [comment]
I've made it to build on all platforms.
I've sent you an email about this, @jdm :)

**Comment by jdm** [comment]
It looks like that rebase didn't quite work correctly.

**Comment by ecoal95** [comment]
Yep, I know, I'm reopening it, sorry.

**Pull request closed without merging**